File: discord-bot-fleet/src/store.ts
// Persistent config store. Single source of truth: data/bots.json
// All writes go through this module so we can encrypt/validate atomically.

import { existsSync, mkdirSync, readFileSync, writeFileSync, renameSync } from 'fs';
import { join } from 'path';
import { nanoid } from 'nanoid';
import type { BotConfig } from './types.js';

const DATA_DIR = process.env.DATA_DIR || join(process.cwd(), 'data');
const BOTS_FILE = join(DATA_DIR, 'bots.json');

interface StoreShape {
  bots: BotConfig[];
}

function ensureDataDir() {
  if (!existsSync(DATA_DIR)) mkdirSync(DATA_DIR, { recursive: true });
}

function load(): StoreShape {
  ensureDataDir();
  if (!existsSync(BOTS_FILE)) {
    writeFileSync(BOTS_FILE, JSON.stringify({ bots: [] }, null, 2));
    return { bots: [] };
  }
  try {
    const raw = readFileSync(BOTS_FILE, 'utf8');
    const parsed = JSON.parse(raw);
    if (!parsed.bots || !Array.isArray(parsed.bots)) throw new Error('invalid bots.json');
    return parsed as StoreShape;
  } catch (e) {
    console.error('[store] failed to parse bots.json, backing up and starting fresh:', e);
    renameSync(BOTS_FILE, BOTS_FILE + '.bak.' + Date.now());
    writeFileSync(BOTS_FILE, JSON.stringify({ bots: [] }, null, 2));
    return { bots: [] };
  }
}

function save(store: StoreShape) {
  ensureDataDir();
  const tmp = BOTS_FILE + '.tmp';
  writeFileSync(tmp, JSON.stringify(store, null, 2));
  renameSync(tmp, BOTS_FILE); // atomic rename
}

export function listBots(): BotConfig[] {
  return load().bots;
}

export function getBot(id: string): BotConfig | undefined {
  return load().bots.find(b => b.id === id);
}

export function createBot(input: Partial<BotConfig> & { name: string; persona: string; token_enc: string; llm: BotConfig['llm']; gating: BotConfig['gating']; tools: BotConfig['tools'] }): BotConfig {
  const store = load();
  const now = Date.now();
  const bot: BotConfig = {
    id: input.id || nanoid(10),
    name: input.name,
    persona: input.persona,
    token_enc: input.token_enc,
    guild_id: input.guild_id || '',
    channel_ids: input.channel_ids || [],
    delegated_bots: input.delegated_bots || [],
    discord_user_id: input.discord_user_id,
    llm: input.llm,
    gating: input.gating,
    tools: input.tools,
    status: input.status || 'stopped',
    created_at: now,
    updated_at: now,
  };
  store.bots.push(bot);
  save(store);
  return bot;
}

export function updateBot(id: string, patch: Partial<BotConfig>): BotConfig {
  const store = load();
  const idx = store.bots.findIndex(b => b.id === id);
  if (idx === -1) throw new Error(`bot ${id} not found`);
  const updated: BotConfig = {
    ...store.bots[idx],
    ...patch,
    id, // can't change id
    updated_at: Date.now(),
  };
  store.bots[idx] = updated;
  save(store);
  return updated;
}

export function deleteBot(id: string): void {
  const store = load();
  store.bots = store.bots.filter(b => b.id !== id);
  save(store);
}

export function setBotStatus(id: string, status: BotConfig['status']) {
  return updateBot(id, { status });
}

File: discord-bot-fleet/bot-python/store.py
"""Config store — reads/writes the same bots.json the Node panel uses.
Uses fcntl file locking to avoid races with the Node process.
"""

import json
import os
import fcntl
import time
from typing import Optional, Dict, Any, List

DATA_DIR = os.environ.get('DATA_DIR', os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'data'))
BOTS_FILE = os.path.join(DATA_DIR, 'bots.json')


def _load() -> Dict[str, Any]:
    os.makedirs(DATA_DIR, exist_ok=True)
    if not os.path.exists(BOTS_FILE):
        with open(BOTS_FILE, 'w') as f:
            json.dump({'bots': []}, f)
        return {'bots': []}
    try:
        with open(BOTS_FILE, 'r') as f:
            fcntl.flock(f.fileno(), fcntl.LOCK_SH)
            data = json.load(f)
            fcntl.flock(f.fileno(), fcntl.LOCK_UN)
        if 'bots' not in data or not isinstance(data['bots'], list):
            raise ValueError('invalid bots.json')
        return data
    except Exception as e:
        # Backup and start fresh
        backup = BOTS_FILE + f'.bak.{int(time.time())}'
        try:
            os.rename(BOTS_FILE, backup)
            print(f'[store] failed to parse bots.json, backed up to {backup}: {e}', flush=True)
        except Exception:
            pass
        with open(BOTS_FILE, 'w') as f:
            json.dump({'bots': []}, f)
        return {'bots': []}


def _save(data: Dict[str, Any]) -> None:
    os.makedirs(DATA_DIR, exist_ok=True)
    tmp = BOTS_FILE + '.tmp'
    with open(tmp, 'w') as f:
        fcntl.flock(f.fileno(), fcntl.LOCK_EX)
        json.dump(data, f, indent=2)
        f.flush()
        os.fsync(f.fileno())
        fcntl.flock(f.fileno(), fcntl.LOCK_UN)
    os.replace(tmp, BOTS_FILE)


def list_bots() -> List[Dict[str, Any]]:
    return _load()['bots']


def get_bot(bot_id: str) -> Optional[Dict[str, Any]]:
    for b in _load()['bots']:
        if b.get('id') == bot_id:
            return b
    return None


def update_bot(bot_id: str, patch: Dict[str, Any]) -> Dict[str, Any]:
    data = _load()
    for i, b in enumerate(data['bots']):
        if b.get('id') == bot_id:
            merged = {**b, **patch, 'id': bot_id, 'updated_at': int(time.time() * 1000)}
            data['bots'][i] = merged
            _save(data)
            return merged
    raise KeyError(f'bot {bot_id} not found')
